# What I did

1. Get mappability of smallest k-mer in BAM file (36-mer).
2. Write script `wig_fixer.awk` to fix the format of mappability.wig files and only include given chromosomes (`wig_fixer.awk`).
3. Write script `separator.awk` which separates the mappability into several files by chromosome (`separator.awk`).
4. Submit MaSC run:
```sh
bsub -M 16384 -R "rusage[mem=16384, scratch=16384]" -n 8 -o masc.out -J MaSC ./MaSC.pl --verbose --mappability_path=../mappability/mappability_separated/ --chrom_length_file=../lengths.chrom.sizes --input_bed=../bam/heart_adult_8_weeks_H3K27me3_ENCODE_854_no_duplicates.bed --prefix=mus_musculus
```
Used ~ 1 GB memory and 1.2 GB swap.
Output: Naive Fragment Length:129, MaSC Fragment Length:137

5. Mask the original wig file with its binary mappability. `wiggletools write mask.wig mult original.wig mappability.wig`
    * Note: `original.wig` was produced by filtering the original bam file as such: `samtools view -b original.bam 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 X Y | wiggletools write original.wig sam -`
6. Shift masked file by fragment length / 2. `wiggletools write shift.wig shiftPos 68 mask.wig`
7. Smooth shifted file via Tukey (or other) kernel. `wiggletools write smooth.wig smooth shift.wig`

# Questions
1. What's better: Calculate fragment size using MaSC (~30 min) or user supplied estimates.
    * MaSC: More accurate. Takes longer. Requires BED file input (via convert2bed), not BAM.
    * Answer: Use Wiggletools to implement self.
2. Why does `align2rawsignal` need a range of k-mers? (Maybe for the cross correlation/related to next question.)
3. Why does the mouse BAM file contain two k-mer lengths (36 bp and 50 bp)?
```
19862515 36 bp
28336996 50 bp
```
4. What are the extra chromosome IDs?
5. How to get Gi (total mappable genome size over both strands)
