# What I did

1. Get mappability of smallest k-mer in BAM file (36-mer).
2. Write script `wig_fixer.awk` to fix the format of mappability.wig files and only include given chromosomes.
`wig_fixer.awk`:
```awk
BEGIN {
    out_file = "2_mus_musculus_all_chrom.map.wig";
    skip = "";
}
{
    tmp = substr($2, 7);
    if ($1 == "fixedStep")
    {
        if (tmp == "1" || tmp == "2" || tmp == "3" || tmp == "4" || tmp == "5" || tmp == "6" || tmp == "7" || tmp == "8" || tmp == "9" || tmp == "10" || tmp == "11" || tmp == "12" || tmp == "13" || tmp == "14" || tmp == "15" || tmp == "16" || tmp == "17" || tmp == "18" || tmp == "19" || tmp == "X" || tmp == "Y")
        {
            print (substr($2, 7)"\t"substr($3, 7)"\t"(substr($3, 7) + substr($4, 6))"\t1") >> out_file;
        }
        skip = "true";
    }
    else if (skip == "true")
    {
        skip = "false";
    }
    else if (tmp == "1" || tmp == "2" || tmp == "3" || tmp == "4" || tmp == "5" || tmp == "6" || tmp == "7" || tmp == "8" || tmp == "9" || tmp == "10" || tmp == "11" || tmp == "12" || tmp == "13" || tmp == "14" || tmp == "15" || tmp == "16" || tmp == "17" || tmp == "18" || tmp == "19" || tmp == "X" || tmp == "Y")
    {
        print $0 >> out_file;
    }
}
```
3. Write script `separator.awk` which separates the mappability into several files by chromosome.
`separator.awk`:
```awk
BEGIN {
    prev = 1;
    output_pre = "./mappability_separated/mus_musculus";
    output_ext = ".map";
    output_file = output_pre"_1"output_ext;
}

{
    if (prev == $1)
    {
        print $0 >> output_file;
    }
    else
    {
        prev = $1;
        output_file = (output_pre"_"$1output_ext);
        print $0 >> output_file;
    }
}
```
4. Submit MaSC run:
```sh
bsub -M 16384 -R "rusage[mem=16384, scratch=16384]" -n 8 -o masc.out -J MaSC ./MaSC.pl --verbose --mappability_path=../mappability/mappability_separated/ --chrom_length_file=../lengths.chrom.sizes --input_bed=../bam/heart_adult_8_weeks_H3K27me3_ENCODE_854_no_duplicates.bed --prefix=mus_musculus
```
Used ~ 1 GB memory and 1.2 GB swap.
Output: Naive Fragment Length:129, MaSC Fragment Length:137

5. Mask the original wig file with its binary mappability. `wiggletools write mask.wig mult original.wig mappability.wig`
* Note: `original.wig` was produced by filtering the original bam file as such: `samtools view -b original.bam 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 X Y | wiggletools write original.wig sam -`
6. Shift masked file by fragment length / 2. `wiggletools write shift.wig shiftPos 68 mask.wig`
7. Smooth shifted file via Tukey (or other) kernel. `wiggletools write smooth.wig smooth shift.wig`

# Questions
1. What's better: Calculate fragment size using MaSC (~30 min) or user supplied estimates.
    * MaSC: More accurate. Takes longer. Requires BED file input (via convert2bed), not BAM.
2. Why does `align2rawsignal` need a range of k-mers?
3. Why does the mouse BAM file contain two k-mer lengths (36 bp and 50 bp)?
```
19862515 36 bp
28336996 50 bp
```
4. What are the extra chromosome IDs?
5. How to get Gi (total mappable genome size over both strands)
