# What I did

1. Get mappability of smallest k-mer in BAM file (36-mer).
2. Write script `wig_fixer.awk` to fix the format of mappability.wig files and only include given chromosomes.
`wig_fixer.awk`:
```awk
BEGIN {
    out_file = "filtered_mus_musculus_all_chrom.map.wig";
}
{
    chr = $1;
    if ($1 == "fixedStep")
    {
        chr = substr($2, 7);
        if (chr == "1" || chr == "2" || chr == "3" || chr == "4" || chr == "5" || chr == "6" || chr == "7" || chr == "8" || chr == "9" || chr == "10" || chr == "11" || chr == "12" || chr == "13" || chr == "14" || chr == "15" || chr == "16" || chr == "17" || chr == "18" || chr == "19" || chr == "X" || chr == "Y")
        {
            prevStart = substr($3, 7);
            prevStep = substr($4, 6);
            prevChr = $1;
            counter = 1;
        }
    }
    else if ($1 == "1.000000")
    {
        print (prevChr"\t"(prevStart + prevStep)"\t"(prevStart + (prevStep * counter))"\t1") >> out_file;
        counter++;
    }
    else if (chr == "1" || chr == "2" || chr == "3" || chr == "4" || chr == "5" || chr == "6" || chr == "7" || chr == "8" || chr == "9" || chr == "10" || chr == "11" || chr == "12" || chr == "13" || chr == "14" || chr == "15" || chr == "16" || chr == "17" || chr == "18" || chr == "19" || chr == "X" || chr == "Y")
    {
        print $0 >> out_file;
    }
}
```
3. Write script `separator.awk` which separates the mappability into several files by chromosome.
`separator.awk`:
```awk
BEGIN {
    prev = 1;
    output_pre = "./mappability_separated/mus_musculus";
    output_ext = ".map";
    output_file = output_pre"_1"output_ext;
}

{
    if (prev == $1)
    {
        print $0 >> output_file;
    }
    else
    {
        prev = $1;
        output_file = (output_pre"_"$1output_ext);
        print $0 >> output_file;
    }
}
```
4. Submit MaSC run:
```sh
bsub -M 16384 -R "rusage[mem=16384, scratch=16384]" -n 8 -o masc.out -J MaSC ./MaSC.pl --verbose --mappability_path=../mappability/mappability_separated/ --chrom_length_file=../lengths.chrom.sizes --input_bed=../bam/heart_adult_8_weeks_H3K27me3_ENCODE_854_no_duplicates.bed --prefix=mus_musculus
```
Used ~ 1 GB memory and 1.2 GB swap.
Output: Naive Fragment Length:129, MaSC Fragment Length:137

5. Mask the original wig file with its binary mappability. `wiggletools write mask.wig mult original.wig mappability.wig`
    * Note: `original.wig` was produced by filtering the original bam file as such: `samtools view -b original.bam 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 X Y | wiggletools write original.wig sam -`
6. Shift masked file by fragment length / 2. `wiggletools write shift.wig shiftPos 68 mask.wig`
7. Smooth shifted file via Tukey (or other) kernel. `wiggletools write smooth.wig smooth shift.wig`

# Questions
1. What's better: Calculate fragment size using MaSC (~30 min) or user supplied estimates.
    * MaSC: More accurate. Takes longer. Requires BED file input (via convert2bed), not BAM.
2. Why does `align2rawsignal` need a range of k-mers?
3. Why does the mouse BAM file contain two k-mer lengths (36 bp and 50 bp)?
```
19862515 36 bp
28336996 50 bp
```
4. What are the extra chromosome IDs?
5. How to get Gi (total mappable genome size over both strands)
