# Comments
1. MaSC.pl outline:
    1. Store chromosome lengths.

    2. For each mappability file (for each chromosome):
        * Create pos and neg vectors of length = chrom length + shift.
        * Put pos and neg vector references into pos and neg hashes, with chromosome identifier as key.
        * Create new vector of length chrom length + shift, set mappable ranges to 1.
        * Store reference of mappability vector into a hash.

    3. Open input file. For each READ:
        * If the read maps to negative strand, then turn on bit of neg vector for that chromosome at END position.
        * If the read maps to positive strand, then turn on bit of pos vector for that chromosome at START position.
        * Else, read maps to unknown chromosome, add to an unknown chromosome hash and increment.
        > `Wiggletools write neg.wig getNeg reads.bam` and `Wiggletools write pos.wig getPos reads.bam`

    4. For each chromosome in neg vector hash:
        * Skip chromosomes with bit vectors all 0 (no reads mapped).
        * Create a neg_map_vector and pos_map_vector which are clones of original mappability vector.
        * Rotate neg_map_vector left by read length (36 most significant bits are put into least significant position).
        > `wiggletools shiftPos READLEN neg_map.wig`

        * Count total reads in neg and pos vectors. (ONLY FOR NAIVE CALCULATIONS)
        * Rotate neg_map_vector and neg_vector right for value set to min_shift.
        > `wiggletools shiftPos -MINSHIFT neg_map` and `wiggletools shiftPos -MINSHIFT neg.wig`

        * For each shift value:
            * Get neg and pos map overlap (neg_map_vector && pos_map_vector), store total doubly mapped read count. (A)
            > `wiggletools write a.wig overlaps pos_map.wig neg_map.wig`

            * Get neg and pos strand overlap (neg_vector && pos_vector), store count. (B)
            > `wiggletools write b.wig overlaps pos.wig neg.wig`

            * Get overlap between above two vectors (positions where reads mapped to pos and neg strand, and are mappable in both strands, A && B), store count. (C)
            > `wiggletools write c.wig overlaps a.wig b.wig`

            * Get neg and double_overlap overlap (reads mapping to negative strand which are mappable in both, neg_vector && A), store count. (D) (For all chromosomes, when neg is shifted by X, this many positions are covered in the negative strand.) (First variable for pearson correlation?)
            > `wiggletools write d.wig overlaps a.wig neg.wig`

            * Get pos and double_overlap overlap (reads mapping to positive strand which are mappable in both, pos_vector && A), store count. (E) (Second variable for pearson correlation?)
            > `wiggletools write e.wig overlaps a.wig pos.wig`

            * Rotate neg_vector and neg_map_vector one to the right (LSB goes to MSB).
            > `wiggletools write neg.wig shiftPos -SHIFT neg.wig` and `wiggletools write neg_map.wig shiftPos -SHIFT neg_map.wig`

    5. Get mean and variance of negative and positive read counts.

    6. For each shift value:
        * Calculate mean doubly mapped reads from positive (E/A) and negative strand (D/A).
        * Calculate variance.
        * Calculate naive coefficient: ((total # neg and pos strand overlap [B] / total genome length) - neg_mean*pos_mean)/sqrt(neg_var*pos_var)
        * Calculate masc coefficient: ((total # neg and pos strand overlap which doubly maps [C]) / total # of doubly mappable reads [A] - pos_mean_mdr*neg_mean_mdr)/sqrt(pos_var_mdr*neg_var_mdr)

    7. Smooth with sliding average, get max correlation values for naive and masc.
2. Create wiggletools getNeg and wiggletools getPos?
3. Translate MaSC to wiggletools:
    * Given: shift value CURSHIFT, mappability file map.wig, reads reads.bam, read length READLEN, minimum shift value MINSHIFT:
        * ```
          for CURSHIFT in MINSHIFT:MAXSHIFT
              wiggletools write neg_pos_map.wig overlaps map.wig shiftPos READLEN - CURSHIFT map.wig
              wiggletools pearson overlaps neg_pos_map.wig unit getPos bam.bam overlaps neg_pos_map.wig shiftPos -CURSHIFT unit getNeg bam.bam
          ```
    * First attempt: READLEN = 36, MINSHIFT = 0, CURSHIFT = 137 Correlation value = -0.594588 w/o unit: -0.206025
    * Second attempt: READLEN = 36, MINSHIFT = 0, CURSHIFT = 3 Correlation value = -0.639954 w/o unit: -0.007922
    * Third attempt: READLEN = 36, MINSHIFT = 0, CURSHIFT = 135 Correlation value = -0.594580 w/o unit: -0.201035
4. Try to create a script for this...
```sh
[joshjayk@ebi-cli-001 wiggletools_masc]$ bsub -M 8192 -R "rusage[mem=8192, scratch=16384]" -n 16 -J accurate ./wiggletools_max_shift.py -c 16 -i 0 -e 400 -s 1 -a -p pos.sorted.wig -n neg.sorted.wig -o accurate_test.out

Job <2518390> is submitted to default queue <research-rh7>.
```
5. Second attempt at translation ????????????????????????????:
    * Given: shift value CURSHIFT, mappability file map.wig, reads reads.bam, read length READLEN, minimum shift value MINSHIFT:
        * ```
          for CURSHIFT in MINSHIFT:MAXSHIFT
              wiggletools write neg_pos_map.wig overlaps map.wig shiftPos CURSHIFT - READLEN map.wig
              wiggletools pearson overlaps neg_pos_map.wig unit getPos bam.bam overlaps neg_pos_map.wig shiftPos CURSHIFT - READLEN unit getNeg bam.bam
          ```

# Questions
1. Why does MaSC rotate neg_pos_map to the left by read length?
    * Answer: Vectors lengths are chromosome length + max_shift - min_shift. So, last bits are all 0 mask.
2. Why does map use rotate instead of shift?
    * Also answered above?
