# Today:
1. Test the pipeline on one file `U937_VD3_LPS_ATAC_ATAC-seq_P2MonocyteMacrophage87_ATAC_P2MonocyteMacrophage87_bwa_samse_BR2TR1_edited.bam`
    * Neg read counts: 97805787
    * Pos read counts: 97813738
    * Sum: 195619525
    * Mappable genome size: 2696858585
    * Ratio: 5.440205303
    * Max threshold: 10.880400 / 4 = 2.7201

2. Generate simulated data.
    * *Question*: Should I be shifting by frag length or frag length / 2???
      Answer: Shift by *shift length* not *fragment length*.

3. Enable multiprocessing in pairwise_pearson.py

4. Resubmit pipeline with corrected normalization shift value.

5. Consider doing it without smooth? Since the ENCODE stuff uses only read starts and not the entire coverage.

6. Fixed read length in normalization python file.

# To do:

1. Rerun on simulation, providing read length manually, with and without smoothing.
