# Today:
1. Fix the mappability for homo sapiens to include chromosomes 20-22.
2. Write preprocess.py script.
3. Download GRCh37, as that's what encode is based off of. Create mappability for that to repeat everything.
    * Awk command to fix chromosomes from last time: `awk 'BEGIN {wrong=1}{if (substr($2,7,6) == "CM0006") {print ($1" chrom=chr"substr($6, 1, length($6) - 1)" "$11);wrong=0} else if (wrong == 0 && $1 ~ /^[0-9]+$/) {print $0} else if ($2 != "chrom=CM000663.2") {wrong=1;}}' GCA_000001405.28_GRCh38.p13_genomic.genmap.wig > fixed_GCA_000001405.28_GRCh38.p13_genomic.genmap.wig`
    * Awk command this time: `awk 'BEGIN {wrong=1}{if (substr($2,7,7) == "NC_0000") {print ($1" chrom=chr"substr($6, 1, length($6) - 1)" "$10);wrong=0} else if (wrong == 0 && $1 ~ /^[0-9]+$/) {print $0} else {wrong=1;}}' GCF_000001405.25_GRCh37.p13_genomic.genmap.wig > fixed.genmap.wig`
4. Realized: I Don't need to run wig_fixer.awk... Just need to run separater. Unused chromosomes don't matter.
5. Ran wiggletools_masc and MaSC for all examples. They both failed for ex1. MaSC succeeded in ex2.
6. Download 4 mouse paired end samples. Run following to unpair them: `awk '{OFS="\t"}{gsub("81|145", "16", $2); gsub("65|129", "0", $2); print $0 >> "unpaired.sam"}' GSM582290_lko100.combined.sam`
7. Convert sam to bam `samtools view -Sbh unpaired.sam > *************.unpaired.bam`
8. Preprocess the four examples.
9. Download mm9 genome, index and calculate mappability for 35 bp reads.
